kubelet/userns: Wrap error message

rata opened the pull request:
Most error messages are properly wrapped already, but this was missing.
While we are there, let's test the wrapping we just added and other errors that are already wrapped.

The third commit just makes sure to use `t.TempDir()` on a test that will write to that directory.

See the individual commits for more details.


<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it: 
This way we show a clear error if the kubelet initialization fails when the feature gate is enabled.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

rata committed:
kubelet/userns: Wrap error message

Most error messages are properly wrapped already, but this was missing.

The kubelet logs will show something like this now:

	E0201 12:00:03.505680 3007049 run.go:74] "command failed" err="failed to run Kubelet: failed to create kubelet: record pod mappings: create user namespace store: mkdir XXX: permission denied"

Before this commit, the message was not so clear:

	E0120 16:02:40.484404  474711 run.go:74] "command failed" err="failed to run Kubelet: failed to create kubelet: mkdir XXX: permission denied"

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata committed:
kublet/userns: Test error messages on init failures

This adds a test for the just added wrapping error message, as well as
for the other already present error messages that initialization can
fail with.

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata committed:
kubelet/userns: Use t.TempDir()

These tests will create the userns record mapping file, so let's use a
temporal directory for that.

Without specifying one, by mistake we were using the
"/tmp/non-existant-dir.This-is-not-used-in-tests/" directory.

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata commented:
cc @giuseppe

bart0sh commented on pkg/kubelet/userns/userns_manager.go:
```diff
@@ -157,7 +157,7 @@ func MakeUserNsManager(kl userNsPodsManager) (*UsernsManager, error) {
 		klog.V(5).InfoS("reading pod from disk for user namespace", "podUID", podUID)
 		if err := m.recordPodMappings(podUID); err != nil {
-			return nil, err
+			return nil, fmt.Errorf("user namespace manager can't record pod mappings: %w", err)
```
Would it be better to do this on the lower level, considering that it's already partly done: `recordPodMappings` calls `parseUserNsFileAndRecord`, which returns properly formatted errors.

rata replied:
I don't think so. `parseUserNsFileAndRecord()` is not called unconditionally, it can fail before calling that. I've updated the commit message with examples of how it failed before and how it fails now, so it's clear the improvement.

But I'll put them here too. The kubelet logs will show something like this now:
    
            E0120 16:02:40.484404  474711 run.go:74] "command failed" err="failed to run Kubelet: failed to create kubelet: user namespace manager can't record pod mappings: mkdir XXX: permission denied"
    
Before this commit, the message was not so clear:
    
            E0120 16:02:40.484404  474711 run.go:74] "command failed" err="failed to run Kubelet: failed to create kubelet: mkdir XXX: permission denied"
    
Which is not clear at all, the run.go:74 is in a very generic file that doesn't give you many clues of what is really calling.

Even for the error from `parseUserNsFileAndRecord()`, this is way more clear now.

bart0sh replied:
That's exactly what I mean: if **all** lower level functions called from recordPodMappings would return properly formatted errors, you wouldn't need to wrap them in this function. `parseUserNsFileAndRecord` does this, but it's not the only function called from `recordPodMappings`.

rata replied:
But how would the error printed on kubelet initialization look like? I don't think it will be clear it is the user namespace manager that is failin to do something, right?

I'm open to ideas, please share how the example log would look like, my thinking was that it wasn't more clear than this (for this log line that is printed in a mega generic function, so the line doesn't give you any useful hint), that is why I did it. But maybe I missed something?

bart0sh replied:
Here is what I think about it. We have unwrapped error "mkdir: permission denied" produced by the code in the `pkg/kubelet/util/store/filestore.go:NewFileStore`. This function is called from 3 different places in the k/k code (even from 4 if we consider test code). I think it makes sense to wrap the error on the lowest possible level than wrapping it in 4 different places. However, if something like "NewFileStore: failed to create file store: %w" is not clear enough, we can decide to wrap it again at a higher level. Alternatively we can wrap it in all places where NewFileStore is called and remember to wrap it every time we use NewFileStore.

In my opinion "user namespace manager can't record pod mappings:  mkdir: permission denied" gives more hints than "NewFileStore: failed to create file store: mkdir permissions denied", but this is just my opinion.

rata replied:
The big difference IMHO is that you can grep the userns error and you will find the issue quite easily, while it is not that simple if you only wrap it on the lowest level (it is called from different places, it is not clear where it is called during initialization, etc.). If we wrap it at the userns manager level, we can easily find the issue.

I'm not against adding more wrappers, though, if you have a specific suggestion (i.e. what message to add and where), I can add that too.

rata replied:
@bart0sh friendly ping?

AFAIK, this is not waiting on author as it is marked now

bart0sh replied:
Sorry, missed your answer, but still tend to disagree that "user namespace manager can't record pod mappings: mkdir: permission denied" doesn't give more hints than "NewFileStore: failed to create file store: mkdir permissions denied". From my point of view latter is more logical and understandable.

I'd give other reviewers an opportunity to express their opinions.

rata replied:
@bart0sh I added wrappings for file store too now (and simplified some other error messages along the way). Let me know what you think!

bart0sh replied:
/lgtm

rata committed:
pkg/kubelet/userns: Simplify error messages

The error we are wrapping is already verbose, let's just use minimal
wrapping as it is usually the case in go code.

Note that the error on parseUserNsFileAndRecord() can be returned to the
user, so we added some context about user namespace. Otherwise, an error
to parse the json would not be clear to which of all the json the kubelet
parses it refers to.

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

bart0sh commented:
/triage accepted
/priority important-longterm

/assign @mrunalp @derekwaynecarr @dchen1107 
for a final approval

rata commented:
/retest

rata commented:
This is an unrelated failure. An apiserver unit test is failing, I think this will be fixed soon in master and tests here will pass.
